platform-operator/metricsexporter/metricsexporter_internal.go

```
@@ -0,0 +1,185 @@
+	// Metrics not included in a map can be added to this array for registration
+	allMetrics = populateAllMetricsList(metricsMap)
+	// This map will be automatically populated with all metrics which were not registered correctly
+	// Metrics in this map will be retried periodically
+	failedMetrics = map[prometheus.Collector]int{}
+
+	registry = prometheus.DefaultRegisterer
+)
+
+// This function is used to determine whether a durationTime for a component metric should be set and what the duration time is
+// If the start time is greater than the completion time, the metric will not be set
+// After this check, the function calculates the duration time and tries to set the metric of the component
+// If the component's name is not in the metric map, an error will be raised to prevent a seg fault
+func metricParserHelperFunction(log vzlog.VerrazzanoLogger, componentName string, startTime string, completionTime string, typeofOperation string) {
+	startInSeconds, _ := time.Parse(time.RFC3339, startTime)
+	startInSecondsUnix := startInSeconds.Unix()
+	completionInSeconds, _ := time.Parse(time.RFC3339, completionTime)
+	completionInSecondsUnix := completionInSeconds.Unix()
+	if startInSecondsUnix >= completionInSecondsUnix {
+		return
+	}
+	totalDuration := (completionInSecondsUnix - startInSecondsUnix)
+	_, ok := metricsMap[componentName]
+	if !ok {
+		log.Errorf("Component %s does not have metrics in the metrics map", componentName)
+		return
+	}
+	if typeofOperation == "upgrade" {
+		metricsMap[componentName].LatestUpgradeDuration.Set(float64(totalDuration))
+	}
+	if typeofOperation == "install" {
+		metricsMap[componentName].LatestInstallDuration.Set(float64(totalDuration))
+	}
+
+}
+func registerMetricsHandlersHelper(log *zap.SugaredLogger) error {
+	var errorObserved error
+	for metric, i := range failedMetrics {
+		err := registry.Register(metric)
+		if err != nil {
+			log.Errorf("Failed to register metric index %v for VPO", i)
+			errorObserved = err
```

What happens if you get an error while registering the metric in two or more separate loop iterations? It looks like the first error will be ignored.

---

The error will not be logged explicitly, you are right. This can be switched pretty easily to log the error for all iterations, do you think that would be good?

---

suggest to collect all the failed metrics as you loop through and return them all as an error like "registration failed for metrics [x, y, z]"

---

And only emit a log message from higher on the stack

---

That would be really good, just one thing is that each metric has its own unique error message. Should I not print each metrics' error maybe?

---

the `errors.Wrap` function may provide the functionality needed here to wrap multiple errors. I assuming once you encounter and error, you still want to continue processing the rest of the loop?

---

yes that is the plan

---

Addressed in newest version :)

---

I updated my error handling with Zaid's error fixes.